STIPS/STIP-009.md

```diff
@@ -0,0 +1,1195 @@
+    }
+    _;
+}
+```
+
+> onlyOwner
+
+```solidity
+modifier onlyOwner(ISetToken _setToken) {
+    require(msg.sender == _manager(_setToken).owner(), "Must be owner");
+    _;
+}
+```
+
+> onlyMethodologist
+
+```solidity
+modifier onlyMethodologist(ISetToken _setToken) {
+    require(msg.sender == _manager(_setToken).methodologist(), "Must be methodologist");
+    _;
+}
+```
+
+> onlyOperator
+
+```solidity
+modifier onlyOperator(ISetToken _setToken) {
+    require(_manager(_setToken).operatorAllowlist(msg.sender), "Must be approved operator");
+    _;
+}
+```
+
+> onlyManager
+
+```solidity
+modifier onlyManager(ISetToken _setToken) {
+    require(address(_manager(_setToken)) == msg.sender, "Manager must be sender");
+    _;
+}
+```
+
+### Internal functions
+
+> invokeManager
+
+Invoke call from manager
+
+```solidity
+function invokeManager(ISetToken _setToken, address _module, bytes memory _encoded)
+```
+
++ call *_manager(_setToken).interactManager(_module, _encoded)*
+
+### Abstract Functions
+
+> _manager
+
+Internal function to grab manager of passed SetToken from extensions data structure
+
+```solidity
+function _manager(ISetToken _setToken) internal virtual view returns (IDelegatedManager)
+```
+
+----
+
+> removeExtension
+
+Removes extension
+
+```solidity
+function removeExtension(ISetToken _setToken) external virtual;
+```
+
+### TradeExtension
+
+#### Inheritance
+
+- BaseGlobalExtension
+
+#### Global Variables
+
+| Type 	| Name 	| Description 	|
+|------	|------	|-------------	|
+|ITradeModule|tradeModule|Trade Module for Set Token|
+
+#### Public Variables
+
+| Type 	| Name 	| Description 	|
+|------	|------	|-------------	|
+|mapping(address => IDelegatedManager)|setManagers|Mapping from Set Token to DelegatedManager|
+
+#### Functions
+
+| Name  | Caller  | Description     |
+|------	|------	|-------------	|
+|initializeExtension|owner|Initialize the TradeExtension on the DelegatedManager|
+|initializeExtensionAndModule|owner|Initialize the StreamingFeeModule on the SetToken and the StreamingFeeExtension on the DelegatedManager|
+|trade|operator|Trade between whitelisted assets on a DEX|
+
+----
+
+### Functions
+
+> initializeExtension
+
+Initialize the TradeExtension on the DelegatedManager
+
+```solidity
+function initializeExtension(address _delegatedManager) external
+```
+
++ require that *msg.sender* is the *_delegatedManager.owner*
++ require that extension state in delegatedManager's *extensionAllowlist* is *PENDING*
++ set *setManagers[_setTokenAddress]* to manager
++ call `manager.initializeExtension()`
++ emit *ExtensionInitialized* event
+
+----
+
+> initializeModuleAndExtension
+
+Initialize the TradeModule and the TradeExtension. (Can only be called by delegatedManager owner, checked in `initializeExtension`)
+
+```
+function initializeModuleAndExtension(address _delegatedManager) external
+```
+
++ call *initializeExtension(_delegatedManager)*
```

Comment:
initializeExtension call needs to happen after module is initialized?

Reply:
Oh... Brian and I talked about this and think we shouldn't impose this condition... 

Do you think there's a reason why the module has to be initialized first?

Reply:
If module is not initialized and the owner calls initializeExtension they will successfully initialize the extension but when they try to use any module functionality it won't work. They'll have to call a separate contract (the module) to initialize.

Reply:
You're right. This my understanding of the sequencing problem (please correct if wrong :) ):
+ to initialize the module we need to call `invokeManager` on the module with the module intialization bytecode
+ invokeManager is `onlyExtension` (which requires that the extension is initialized)
+ to initialize a module we need to initialize the extension first
+ initializeExtension is used in two contexts (one where we know it the module will be initialized second, and one where we hope it's already been initialized). 

So... if that's right, how should we manage the second context?